packages/sync-actions/test/product-sync-variants.spec.js

```diff
@@ -291,42 +291,186 @@ describe('Actions', () => {
-      ],
-    }
+    describe('with variants in `now`', () => {
+      const before = {
+        id: '123',
+        version: 1,
+        variants: [
+          { id: 2, sku: 'v2', key: 'v2', attributes: [{ name: 'foo', value: 'qux' }] },
+          { id: 3, sku: 'v3', key: 'v3', attributes: [{ name: 'foo', value: 'baz' }] },
+        ],
+      }
 
-    const actions = productsSync.buildActions(now, before)
+      const now = {
+        id: '123',
+        variants: [
+          // changed
+          { id: 2, sku: 'v2', key: 'v2', attributes: [{ name: 'foo', value: 'another value' }] },
+          // changed
+          { id: 3, sku: 'v3', key: 'v3', attributes: [{ name: 'foo', value: 'i dont care' }] },
+          // new
+          { sku: 'v4', key: 'v4', attributes: [{ name: 'foo', value: 'yet another' }] },
+        ],
+      }
+
+      it('should generate `addVariant` and `setAttribute` actions',
+      () => {
+        const actions = productsSync.buildActions(now, before)
+
+        expect(actions).toEqual([
+          { action: 'addVariant', sku: 'v4', key: 'v4', attributes: [{ name: 'foo', value: 'yet another' }] },
+          { action: 'setAttribute', variantId: 2, name: 'foo', value: 'another value' },
+          { action: 'setAttribute', variantId: 3, name: 'foo', value: 'i dont care' },
+        ])
+      })
+    })
 
-    expect(actions).toEqual([
-      { action: 'addVariant', sku: 'v4', key: 'v4', attributes: [{ name: 'foo', value: 'yet another' }] },
-      { action: 'setAttribute', variantId: 2, name: 'foo', value: 'another value' },
-      { action: 'setAttribute', variantId: 3, name: 'foo', value: 'i dont care' },
-    ])
+    describe('with changed non master variant in `now`', () => {
+      describe('with non existing variant', () => {
+        const before = {
+          id: '123',
+          version: 1,
+          masterVariant: {
+            id: 1,
+            sku: 'v1',
+            attributes: [{ name: 'foo', value: 'bar' }],
+          },
+          variants: [],
+        }
+
+        const now = {
+          id: '123',
+          version: 1,
+          masterVariant: {
+            id: 2,
+            sku: 'v1',
+            attributes: [{ name: 'foo', value: 'bar' }],
+          },
+          variants: [],
+        }
+
+        it('should generate `changeMasterVariant` and `addVariant` action',
+        () => {
+          const actions = productsSync.buildActions(now, before)
+
+          expect(actions).toEqual([
+            { action: 'addVariant', attributes: [{ name: 'foo', value: 'bar' }], id: 2, sku: 'v1' },
+            { action: 'changeMasterVariant', variantId: 2 },
+          ])
+        })
+      })
+
+      describe('with existing variant', () => {
+        describe('without changes to attributes', () => {
+          const before = {
+            id: '123',
+            version: 1,
+            masterVariant: {
+              id: 2,
```

Comment:
`masterVariant` cannot have a `variantId` of 2. I think it's always 1

Cc @yanns

Reply:
It can have an id of 2 otherwise you'd never be able to move the master variant. The id is immutable and sequentially assigned.

Reply:
Oh, okay, good to know that `variantId` does not change.

Reply:
I confirm when @tdeekens says.
Master variant id is not always 1. The variant id is the invariant :rofl: